### PR TITLE
Crowbar floor tiles and placement

### PR DESCRIPTION
**Pull request by Ephememory**
Fixes: #371
image

If no tile to drop is set, the tile will drop the generic steel tile at the moment.
- Adds tile removing behavior to **CrowbarComponent**.
- Add **FloorTileItemComponent**.
- Add **genhit.ogg**
- _Add floor tile item prototypes to YAML._
- _Specify which prototype to drop for each **ContentTileDefinition** in YAML._

### CODE REVIEW

File: Content.Server/GameObjects/Components/Interactable/Tools/CrowbarComponent.cs [hunk 1]
```diff
@@ -43,6 +43,10 @@ public void AfterAttack(AfterAttackEventArgs eventArgs)
                 var underplating = _tileDefinitionManager["underplating"];
                 mapGrid.SetTile(eventArgs.ClickLocation, new Tile(underplating.TileId));
                _entitySystemManager.GetEntitySystem<AudioSystem>().Play("/Audio/items/crowbar.ogg", Owner);
+                //Actually spawn the relevant tile item at the right position and give it some offset to the corner.
+                var tileItem = Owner.EntityManager.SpawnEntity(tileDef.ItemDropPrototypeName);
+                tileItem.Transform.GridPosition = coordinates;
+                tileItem.Transform.WorldPosition += new Robust.Shared.Maths.Vector2(0.2f, 0.2f);
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Interactable/Tools/CrowbarComponent.cs]: You can just do this here:

```suggestion
                tileItem.Transform.WorldPosition += (0.2f, 0.2f);
```

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    public class FloorTileItemComponent : Component, IAfterAttack, IExamine
+    {
+#pragma warning disable 649
+        [Dependency] private readonly ITileDefinitionManager _tileDefinitionManager;
+        [Dependency] private readonly IEntitySystemManager _entitySystemManager;
+        [Dependency] private readonly IMapManager _mapManager;
+#pragma warning restore 649
+
+        public override string Name => "FloorTile";
+        public int CurrentStackCount = 1;
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: There is already a `StackComponent` that you can use to handle the whole stacking business.

Comment (Ephememory) [reply Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: 🤦‍♀

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs [hunk 3]
```diff
@@ -66,27 +53,20 @@ public void AfterAttack(AfterAttackEventArgs eventArgs)
...
-            if (tileDef.IsSubFloor)
+            if (tileDef.IsSubFloor && attacked == null)
             {
                 var desiredTile = _tileDefinitionManager[_outputTile];
                 mapGrid.SetTile(eventArgs.ClickLocation, new Tile(desiredTile.TileId));
                 _entitySystemManager.GetEntitySystem<AudioSystem>().Play("/Audio/items/genhit.ogg", Owner);
-
-                CurrentStackCount--;
-                if (CurrentStackCount <= 0)
+                if (!Stack.Use(1))
```
Comment (PJB3005) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: This won't actually do what you want it to. `StackComponent.Use` returns `false` if it *couldn't* remove from the stack anymore.

You'll need to manually check for count 0 after the use and delete then.

Comment (RemieRichards) [reply Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: Also, due to the placement of the if, a 0-size floortile will still place a tile down.
Granted we shouldn't be getting 0-size tiles but hey, it happened in SS13...

Comment (Ephememory) [reply Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: Thankya

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            var coordinates = mapGrid.GridTileToLocal(tile.GridIndices);
+            float distance = coordinates.Distance(_mapManager, Owner.Transform.GridPosition);
+
+            if (distance > InteractionSystem.InteractionRange)
+            {
+                return;
+            }
+
+            var tileDef = (ContentTileDefinition)_tileDefinitionManager[tile.Tile.TypeId];
+            if (tileDef.IsSubFloor && attacked == null)
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: The issue with 0-count tiles making a tile is still present.
You need to move the ```Stack.Use(1)``` up into this if

```suggestion
            if (tileDef.IsSubFloor && attacked == null && Stack.Use(1))
```

Just moving it into the if (removing it from where it is now) will solve it.

File: Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            {
+                var desiredTile = _tileDefinitionManager[_outputTile];
+                mapGrid.SetTile(eventArgs.ClickLocation, new Tile(desiredTile.TileId));
+                _entitySystemManager.GetEntitySystem<AudioSystem>().Play("/Audio/items/genhit.ogg", Owner);
+                Stack.Use(1);
```
Comment (RemieRichards) [inline Content.Server/GameObjects/Components/Items/FloorTileItemComponent.cs]: See above
```suggestion
```

**Pull request merged**